---
title: "Analyze host called start sites"
author: "`r Sys.info()[['user']]`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  github_document:
    toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
```

## Import libraries
```{r}
library(Biostrings)
library(GenomicRanges)
library(GenomicFeatures)
library(BSgenome)
library(broom)
library(biobroom)
library(glue)
library(stringr)
library(grid)
library(plyranges)
library(tidyverse)
library(rasilabRtemplates)

sample_names <- c(
  "vir" = "+vir",
  "ifn_vir" = "+ifn +vir",
  "untr" = "ctrl",
  "ifn" = "+ifn"
)

# use this for finding unique sample combinations
sample_numbers <- c(
  "vir" = 8,
  "ifn_vir" = 2,
  "untr" = 4,
  "ifn" = 1
)


treatment_names <- c(
  "cyclo" = "ribo",
  "ltm" = "ribo + LTM",
  "mrna" = "mRNA"
)
# use this order to resolve multiple TIS within the same 5nt window
tis_priority <- seq(11)
names(tis_priority) <- c("AUG", "CUG", "GUG", "UUG", "AUA", "AUC", "AUU",
                     "AAG", "ACG", "AGG", "other")

```

## Read genome and canonical transcript annotations
```{r}
genome <- BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
annotations <- glue("/fh/fast/subramaniam_a/db/rasi/genomes/human/hg38/gencode/annotations/gencode.v24.canonical_ccds_transcripts.20170315.gff3.gz")
annotations <- rtracklayer::import.gff3(annotations)
```

## Extract transcript annotations and sequence
```{r}
tx <- annotations %>% 
  subset(type == "exon") %>% 
  split(.$transcript_id)

tx_annotations <- annotations %>% 
  subset(type == "transcript") %>% 
  tidy()

tx_seq <- extractTranscriptSeqs(genome, tx) %>% 
  print()
```

## Read in annotated start codons and convert to transcript coords
```{r}
starts <- annotations %>% 
  subset(type == "start_codon") %>% 
  # consider only the first contiguous stretch of each transcript
  tidy() %>%
  group_by(transcript_id) %>%
  filter(start == if_else(strand == "+", min(start), max(start))) %>%
  ungroup() %>%
  GRanges() %>%
  # resize to +1 nt since that is the reference point
  # note that this resize automatically accounts for strand
  resize(width = 1)

starts_tx <- starts %>% 
  # convert to transcript coords
  GenomicFeatures::mapToTranscripts(tx)

starts_tx$transcript_id <- starts$transcript_id[starts_tx$xHits]

# since each start codon is already annotated with the tx it belongs to,
# we can safely ignore the other transcripts that it maps to.
starts_tx <- starts_tx %>% 
  subset(seqnames(.) == transcript_id)

starts_tx <- starts_tx %>% 
  tidy() %>% 
  rename(canonical.start = start, canonical.end = end) %>% 
  select(transcript_id, canonical.start, canonical.end) %>% 
  print()
```

## Read in called start sites
```{r}
cvg <- list.files("../tables/", pattern = "raw_list_host_called_tis.*tsv.gz",
                  full.names = T) %>% 
  map(read_tsv) %>% 
  bind_rows() %>% 
  select(-gene_id, -longest_cds_length) %>% 
  # create a serial number column for joing later
  mutate(xHits = seq(n())) %>% 
  GRanges() %>% 
  print()

cvg_df <- cvg %>% 
  tidy() %>% 
  rename(chr = seqname, chr.start = start, chr.end = end) %>% 
  select(-strand, -width)
```

## Convert called starts to transcriptomic coords from genomic coords
```{r}
cvg_tx <- cvg %>% 
  GenomicFeatures::mapToTranscripts(tx) %>% 
  mutate(strand = "*") %>% 
  # set names to be transcript name
  setNames(seqnames(.)) %>% 
  mutate(tx.hit = seq(length(.))) %>% 
  # extend 1nt in 5' side and 4nt in the 3' side (1nt on either side of codon)
  promoters(upstream = 1, downstream = 4) %>%
  # extract flanking sequence
  mutate(seq = getSeq(tx_seq, .)) %>% 
  print()
```

## Find near cognate matches in a window of 5nt around each called start
```{r}
near_cognate_matches <- vmatchPattern("ATG", cvg_tx$seq, max.mismatch = 1) %>% 
  as("CompressedIRangesList") %>% 
  as_tibble() %>% 
  rename(tx.hit = group) %>%
  filter(start %in% seq(1,3)) %>%
  print()
```

## Extract the near cognate codon from the match
```{r}
near_cognate_match_codon <- subseq(
  x = cvg_tx$seq[near_cognate_matches[['tx.hit']]],
  start = near_cognate_matches[['start']],
  end = near_cognate_matches[['end']]) %>% 
  as.character() %>% 
  enframe("sno", "codon") %>%
  bind_cols(near_cognate_matches) %>% 
  select(codon) %>% 
  print()
```

## Combine the near cognate match location and codon
```{r}
near_cognate_hits <- near_cognate_matches %>% 
  bind_cols(near_cognate_match_codon) %>% 
  mutate(codon.start = start) %>% 
  select(tx.hit, codon.start, codon) %>% 
  print()
```

## Join called starts with assigned near cognate codon and their location
```{r}
cvg_tx_with_canonical <- cvg_tx %>%
  tidy() %>% 
  left_join(near_cognate_hits, by = "tx.hit") %>% 
  left_join(starts_tx, by = c("seqname" = "transcript_id")) %>% 
  mutate(codon = if_else(is.na(codon), "other", codon)) %>% 
  # convert T to U in codons for plotting
  mutate(codon = str_replace_all(codon, "T", "U")) %>% 
  mutate(tis.start = start + codon.start - 1) %>%  
  ## account for non near cognate matches
  mutate(nt_to_atis = if_else(!is.na(tis.start), 
                                          tis.start - canonical.start,
                                          as.numeric(start - canonical.start))) %>% 
  mutate(frame = nt_to_atis %% 3) %>% 
  # if there are multiple near cognates in the 5nt window, get a single one
  # based on the priority order above
  mutate(codon_priority = tis_priority[codon]) %>% 
  # get sample and chromosome locations
  left_join(cvg_df, by = "xHits") %>% 
  group_by(sample, chr, seqname, start) %>% 
  arrange(codon_priority) %>% 
  slice(1) %>% 
  mutate(tis.chr.start = chr.start + codon.start - 1) %>%  
  ungroup() %>%
  select(-canonical.end, -codon.start, -strand, -chr.end, -end, -seq,
         -codon_priority, -tx.hit,  -matches("width|names|Hits")) %>% 
  print()
```

## Get peptide and peptide length of TIS, write all TIS annotations to  file
```{r}
# extract the transcript sequence for each called start
tx_seq_called_starts <- tx_seq[cvg_tx_with_canonical[['seqname']]]
# truncate the sequence starting from the called start to end of transcript
truncated <- subseq(tx_seq_called_starts, cvg_tx_with_canonical[['tis.start']], width(tx_seq_called_starts))
# extract the orf upto the first stop codon
orfs <- str_extract(truncated, "^([ACTG]{3})+?(TAG|TGA|TAA)")
# set all non-orfs and non-near cognate start orfs to blank
orfs[is.na(orfs)] <- ""
orfs[is.na(cvg_tx_with_canonical[['tis.start']])] <- ""
# extract peptides
peptides <- translate(DNAStringSet(orfs))
peptide_length <- width(peptides)
# truncate peptides to max length of 30  for ease of viewing
peptides = subseq(peptides, NA, ifelse(width(peptides) > 30, 30, width(peptides)))

tis <- cvg_tx_with_canonical %>% 
  mutate(peplen = peptide_length, pep = as.character(peptides)) %>% 
  rename(ltm_p = ltm_local_ztnb_pvalue, cyclo_p = cyclo_local_ztnb_pvalue,
         cyclo_n = cyclo_score, ltm_n = ltm_score) %>% 
  select(-matches("threshold"), -seqname, -start) %>% 
  # this gets rid of a single misannotated tis
  filter(!is.na(nt_to_atis)) %>% 
  mutate (tis.type = case_when(
    nt_to_atis < 0 ~ "uTIS", 
    nt_to_atis == 0 ~ "aTIS", 
    nt_to_atis > 0 ~ "dTIS")) %>% 
  write_tsv("../tables/called_host_tis.tsv") %>% 
  print()
```

## Plot the proportion of tis.type within each sample
```{r, fig.width=3.2, fig.height=1.6}
plot_data <- tis %>% 
  group_by(sample, tis.type) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  group_by(sample) %>% 
  mutate(n_prop = n / sum(n) * 100) %>% 
  ungroup() %>% 
  mutate(sample = sample_names[sample])

plot_data %>% 
  ggplot(aes(y = n_prop, x = sample, fill = tis.type)) +
  geom_col() +
  geom_text(data = plot_data %>%
              group_by(sample) %>%
              mutate(n = paste0("N=", sum(n))) %>% slice(1), y = 95, aes(label = n), size = 2.8) +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "sample", y = "TIS proportion (%)", 
       fill = "") +
  guides(fill = guide_legend(keyheight = 0.8)) +
  theme(legend.position = "right")

ggsave('../figures/host_tis_type_proportion.pdf', 
       device = cairo_pdf)
```

## Test if uTIS or dTIS initiation is overrepresented under different treatments

### Get the number of TIS types in each sample
```{r}
test_data <- plot_data %>% 
  select(sample, tis.type, n) %>% 
  spread(tis.type, n, fill = 0) %>% 
  data.table::data.table() %>% 
  data.table::setkey(sample)

print(test_data)
```

### Test  for increased uTIS upon +ifn  treatment
```{r}
data <- test_data[c("+ifn", "ctrl"),c("aTIS", "uTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for increased dTIS upon +ifn treatment
```{r}
data <- test_data[c("+ifn", "ctrl"),c("aTIS", "dTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for increased uTIS upon +ifn +vir treatment
```{r}
data <- test_data[c("+ifn +vir", "ctrl"),c("aTIS", "uTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for increased dTIS upon +ifn +vir treatment
```{r}
data <- test_data[c("+ifn +vir", "ctrl"),c("aTIS", "dTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for increased uTIS upon +vir treatment
```{r}
data <- test_data[c("+vir", "ctrl"),c("aTIS", "uTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for increased dTIS upon +vir treatment
```{r}
data <- test_data[c("+vir", "ctrl"),c("aTIS", "dTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

## Plot number of near-cognates across all samples
```{r, fig.width=3, fig.height=2}
tis %>% 
  group_by(sample, codon) %>% 
  summarize(n_start = n()) %>% 
  mutate(codon_priority = tis_priority[codon]) %>% 
  ungroup() %>% 
  mutate(codon = forcats::fct_reorder(factor(codon), codon_priority)) %>% 
  mutate(sample = sample_names[sample]) %>% 
  ggplot(aes(y = n_start, x = sample, fill = codon)) +
  geom_col() +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "sample", y = "TIS count",
       fill = "TIS identity") +
  guides(fill = guide_legend(keyheight = 0.8))

ggsave('../figures/host_near_cognate_count.pdf', 
       device = cairo_pdf)
```

## Same plot as above but normalized within each sample
```{r, fig.width=3.6, fig.height=2}
plot_data <- tis %>% 
  group_by(sample, codon) %>% 
  summarize(n = n()) %>% 
  mutate(codon_priority = tis_priority[codon]) %>% 
  ungroup() %>% 
  group_by(sample) %>% 
  mutate(n_prop = n / sum(n) * 100) %>% 
  ungroup() %>% 
  mutate(codon = forcats::fct_reorder(factor(codon), codon_priority)) %>% 
  mutate(sample = sample_names[sample])

plot_data %>% 
  ggplot(aes(y = n_prop, x = sample, fill = codon)) +
  geom_col() +
  geom_text(data = plot_data %>%
              group_by(sample) %>%
              mutate(n = paste0("N=", sum(n))) %>% slice(1), y = 95, aes(label = n), size = 2.8) +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "sample", y = "TIS proportion (%)", 
       fill = "TIS identity") +
  guides(fill = guide_legend(keyheight = 0.8))

ggsave('../figures/host_near_cognate_proportion.pdf', 
       device = cairo_pdf)
```

## Prepare table to test whether there are more nearATG codons in the treatment samples compared to control sample
```{r}
test_data <- plot_data %>% 
  filter(codon != "other") %>% 
  mutate(atg = if_else(codon == "AUG", "atg", "nearatg")) %>% 
  group_by(sample, atg) %>%
  summarize(n = sum(n)) %>%
  ungroup() %>%
  # unite(sample, sample, atg, remove = F) %>% 
  spread(atg, n) %>%
  data.table::data.table() %>% 
  data.table::setkey(sample)
print(test_data)
```

Note that the the single-sided "less" test which is appropriate here, but the two-sided test also gives non-significant results

### Test  for +ifn
```{r}
data <- test_data[c("+ifn", "ctrl"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

### Test  for +ifn +vir
```{r}
data <- test_data[c("+ifn +vir", "ctrl"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```


### Test  for +vir
```{r}
data <- test_data[c("+vir", "ctrl"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

## Plot how many near-cognate starts overlap between different samples
```{r, fig.width=3.5, fig.height=3.5}
plot_data <- tis %>% 
  # get rid of peaks that coincide with the same start on canonical tx
  # this occurs because the peaks were called using the longest tx not canonical
  # these starts are spread over multiple exons
  distinct(gene_name, nt_to_atis, codon, sample) %>%
  # get a unique number, name, number of samples in each combination of samples 
  group_by(gene_name, nt_to_atis, codon) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo_num = sum(sample_numbers[sample])) %>% 
  mutate(combo = paste0(sample, collapse = ":")) %>% 
  ungroup() %>% 
  select(-sample) %>% 
  distinct() %>% 
  # count number of tis called each sample combo
  group_by(combo, combo_num, nsamples) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  mutate(sample = map(combo, function(x) str_split(x, ":")[[1]])) %>%
  unnest(sample) %>% 
  select(-combo) %>% 
  mutate(is_called = T) %>% 
  complete(sample, nesting(n, combo_num, nsamples), fill = list("is_called"  = F)) %>% 
  mutate(combination = as.factor(nsamples * 100 + combo_num)) %>% 
  mutate(sample = sample_names[sample]) %>% 
  print()

p <- plot_data %>%
  ggplot(aes(x = sample, y = combination, fill = is_called)) +
  geom_tile(color = "black") +

  viridis::scale_fill_viridis(discrete = T) +
  labs(y = "number of called TIS\n\n", fill = "is_called",
       title = "Overlap in called TIS between samples"
       ) +
  theme(
    panel.spacing = unit(2, "lines"),
    axis.text.y = element_blank(),
    axis.ticks = element_blank(),
    axis.line = element_blank(),
    axis.text.x = element_text(angle = 0, hjust = 0.5)) +
  geom_text(data = plot_data %>% filter(sample == "+ifn"),
            aes(x = 0.4, label = n), hjust = 1, size = 3) +
  scale_x_discrete()

# Code to override clipping
gt <- ggplot_gtable(ggplot_build(p))
gt$layout$clip <- "off"
grid.draw(gt)


ggsave('../figures/host_near_cognate_overlap_between_samples.pdf',
       plot = gt)
# plot(gt)
```


## Same plot as above but faceted by the identity of the start site
```{r,fig.width=10, fig.height=8}
plot_data <- tis %>% 
  # get rid of peaks that coincide with the same start on canonical tx
  # this occurs because the peaks were called using the longest tx not canonical
  # these starts are spread over multiple exons
  distinct(gene_name, nt_to_atis, codon, sample) %>%
  # get a unique number, name, number of samples in each combination of samples 
  group_by(gene_name, nt_to_atis, codon) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo_num = sum(sample_numbers[sample])) %>% 
  mutate(combo = paste0(sample, collapse = ":")) %>% 
  ungroup() %>% 
  select(-sample) %>% 
  distinct() %>% 
  # count number of tis of each type called each sample combo
  group_by(combo, combo_num, nsamples, codon) %>% 
  summarize(n = n()) %>% 
  ungroup() %>% 
  mutate(sample = map(combo, function(x) str_split(x, ":")[[1]])) %>%
  unnest(sample) %>% 
  select(-combo) %>% 
  mutate(is_called = T) %>% 
  complete(sample, nesting(n, combo_num, nsamples, codon), fill = list("is_called"  = F)) %>% 
  mutate(combination = as.factor(nsamples * 100 + combo_num)) %>% 
  mutate(sample = sample_names[sample]) %>% 
  mutate(codon = forcats::fct_reorder(codon, tis_priority[codon], .desc = F)) %>% 
  print()
  
p <- plot_data %>% 
  ggplot(aes(x = sample, y = combination, fill = is_called)) +
  geom_tile(color = "black") +
  facet_wrap(~ codon, ncol = 4, scales = "free") +

  viridis::scale_fill_viridis(discrete = T) + 
  labs(y = "number of called TIS\n\n", fill = "is_called",
       title = "Overlap in called TIS between samples, faceted by codon"
       ) +
  theme(
    panel.spacing = unit(2, "lines"),
    axis.text.y = element_blank(),
    axis.ticks = element_blank(), 
    axis.line = element_blank(),
    axis.text.x = element_text(angle = 0, hjust = 0.5)) +
  geom_text(data = plot_data %>% filter(sample == "+ifn"), 
            aes(x = 0.4, label = n), hjust = 1, size = 3) +
  scale_x_discrete() 

# Code to override clipping
gt <- ggplot_gtable(ggplot_build(p))
gt$layout$clip <- "off"
grid.draw(gt)

ggsave('../figures/host_near_cognate_overlap_between_samples_faceted_by_codon.pdf', 
       plot = gt, device = cairo_pdf)
# plot(gt)
```
