---
title: "Statistics of called host start sites"
author: "`r Sys.info()[['user']]`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  github_document:
    toc: yes
---

In this notebook, we look at different statistics of host called start sites such
as the proportion of uTIS/dTIS vs aTIS and non-AUG vs AUG TIS.

We do these analysis for all genes as well as the genes that are induced upon
treatment with interferon, influenza virus, or both.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

# Import libraries and define analysis specific variables
```{r}
library(tidyverse)
library(rasilabRtemplates)

# use this for labeling axis ticks
sample_names <- c(
  "vir" = "+vir",
  "ifn_vir" = "+ifn +vir",
  "untr" = "ctrl",
  "ifn" = "+ifn"
)

# use this for finding unique sample combinations
sample_numbers <- c(
  "vir" = 8,
  "ifn_vir" = 2,
  "untr" = 4,
  "ifn" = 1
)

# use this order to resolve multiple TIS within the same 5nt window and
# also to order codons during plotting
tis_priority <- seq(11)
names(tis_priority) <- c("AUG", "CUG", "GUG", "UUG", "AUA", "AUC", "AUU",
                     "AAG", "ACG", "AGG", "other")
```

# Read in called starts from our data and analysis of Lee PNAS 2012 data
```{r}
tis <- read_tsv("../tables/called_host_tis.tsv") %>% 
  mutate_at(vars(ends_with("_p")), funs(as.numeric)) %>% 
  mutate(sno = seq(1, n())) %>% 
  print()

lee_2012_tis <- read_tsv("../tables/lee_2012_called_host_tis.tsv") %>%  
  mutate_at(vars(ends_with("_p")), funs(as.numeric)) %>% 
  mutate(sno = seq(1, n())) %>% 
  # this gets rid of a single misannotated tis
  filter(!is.na(nt_to_atis)) %>% 
  mutate (tis.type = case_when(
    nt_to_atis < 0 ~ "uTIS", 
    nt_to_atis == 0 ~ "aTIS", 
    nt_to_atis > 0 ~ "dTIS")) %>% 
  print()
```


# Get combination of samples that each start occurs in
```{r}
tis_combo <- tis %>% 
  # get rid of peaks that coincide with the same start on canonical tx
  # this occurs because the peaks were called using the longest tx not canonical
  # these starts are spread over multiple exons
  distinct(gene_name, nt_to_atis, codon, sample) %>%
  # get a unique number for each combination of samples 
  group_by(gene_name, nt_to_atis, codon) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo_num = sum(sample_numbers[sample])) %>% 
  mutate(combo = paste0(sample, collapse = ":")) %>% 
  ungroup() %>% 
  select(-sample) %>% 
  distinct() %>% 
  left_join(tis %>% select(gene_name, codon, nt_to_atis, tis.type,
                           frame, peplen, pep) %>% distinct(),
            by = c("gene_name", "nt_to_atis", "codon")) %>%
  print()
```

# Create combo with chromosome annotations for joining with unfiltered starts later
```{r}
tis_combo_with_chr <- tis_combo %>% 
  right_join(tis %>% select(codon, nt_to_atis, gene_name, sample, chr, chr.start),
             by = c("codon", "nt_to_atis", "gene_name")) %>% 
  print()
```

# Get combination of samples that each start occurs in Lee 2012
```{r}
lee_2012_tis_combo <- lee_2012_tis %>% 
  # get rid of peaks that coincide with the same start on canonical tx
  # this occurs because the peaks were called using the longest tx not canonical
  # these starts are spread over multiple exons
  distinct(gene_name, nt_to_atis, codon, sample) %>%
  # get a unique number for each combination of samples 
  group_by(gene_name, nt_to_atis, codon) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo_num = sum(sample)) %>% 
  mutate(combo = paste0(sample, collapse = ":")) %>% 
  ungroup() %>% 
  select(-sample) %>% 
  distinct() %>% 
  left_join(tis %>% select(gene_name, codon, nt_to_atis, tis.type,
                           frame, peplen, pep) %>% distinct(),
            by = c("gene_name", "nt_to_atis", "codon")) %>%
  print()
```


# Plot the proportion of TIS that are called in certain number of samples 
```{r, fig.width=1.6, fig.height=1.2}
tis_combo %>% 
  group_by(nsamples) %>% 
  summarize(n = n()) %>% 
  ggplot(aes(x = nsamples, y = n)) +
  geom_col(fill = "grey", color = "black") +
  labs(x = "Number of samples", y = "TIS counts") +
  scale_x_reverse()

ggsave('../figures/tis_vs_nsamples_host.pdf', 
       units = 'in')
```

# Distinct sample combinations for called TIS
```{r}
distinct(tis_combo, combo, n, combo_num) %>% 
  arrange(desc(combo_num)) %>% 
  knitr::kable()
```

# TIS common to all samples
```{r}
common_starts <- tis_combo %>% 
  filter(nsamples == 4) %>% 
  print()
```

# Plot overlap in called TIS between our high cofidence set and Lee 2012 high confidence set

```{r, fig.width=2, fig.height=2}
list1 <- common_starts %>% 
  select(codon, nt_to_atis, gene_name) %>% 
  mutate(sample = "this")

list2 <- lee_2012_tis_combo %>% 
  filter(nsamples == 2) %>% 
  select(codon, nt_to_atis, gene_name) %>% 
  mutate(sample = "lee")

tis_overlap_with_lee <- bind_rows(list1, list2) %>% 
  group_by(codon, nt_to_atis, gene_name) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo = paste0(as.character(sample), collapse = ":")) %>% 
  ungroup() %>% 
  select(-sample) %>% 
  distinct() %>% 
  group_by(combo) %>% 
  dplyr::count() %>% 
  ungroup() %>% 
  as.data.frame() %>% 
  column_to_rownames("combo") %>% 
  print()
  

p <- VennDiagram::draw.pairwise.venn(
  tis_overlap_with_lee[['this', 'n']] + tis_overlap_with_lee[['this:lee', 'n']] ,
  tis_overlap_with_lee[['lee', 'n']] + tis_overlap_with_lee[['this:lee', 'n']], 
  tis_overlap_with_lee[['this:lee', 'n']], 
  category = c("This study", "Lee 2012"), 
  fill = "grey", 
  cat.pos = c(-40, 30),
  cat.dist = c(0.04,0.04))

ggsave("../figures/overlap_between_us_lee_2012.pdf", plot = p)

```

# Proportion of uTIS, aTIS, dTIS among common_starts
```{r, fig.width=1.6, fig.height=1.2}
plot_data <- common_starts %>% 
  group_by(tis.type) %>% 
  summarize(n = n()) %>% 
  print()

plot_data %>% 
  ggplot(aes(x = tis.type, y = n)) +
  geom_col(fill = "grey", color = "black") +
  labs(x = "TIS type", y = "TIS counts") 

ggsave('../figures/atis_vs_dtis_count_host.pdf', 
       units = 'in')

plot_data <- common_starts %>% 
  group_by(tis.type) %>% 
  summarize(n = n()) %>% 
  mutate(n = n / sum(n) * 100) %>% 
  print()

plot_data %>% 
  ggplot(aes(x = tis.type, y = n)) +
  geom_col(fill = "grey", color = "black") +
  labs(x = "TIS type", y = "TIS proportion (%)") 

ggsave('../figures/atis_vs_dtis_proportion_host.pdf', 
       units = 'in')
```
# Proportion of different codons in aTIS and dTIS for common starts

```{r, fig.width=2.8, fig.height=1.6}
plot_data <- common_starts %>% 
  group_by(tis.type, codon) %>% 
  summarize(n = n()) %>% 
  mutate(codon_priority = tis_priority[codon]) %>% 
  ungroup() %>% 
  group_by(tis.type) %>% 
  mutate(n_prop = n / sum(n) * 100) %>% 
  ungroup() %>% 
  mutate(codon = forcats::fct_reorder(factor(codon), codon_priority)) %>% 
  print()
  
plot_data %>%   
  ggplot(aes(y = n_prop, x = tis.type, fill = codon)) +
  geom_col() +
  geom_text(data = plot_data %>% 
              group_by(tis.type) %>% 
              mutate(n = paste0("N=", sum(n))) %>% slice(1), y = 95, aes(label = n), size = 2.8) +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "TIS type", y = "TIS proportion (%)", 
       fill = "") +
  guides(fill = guide_legend(keyheight = 0.8))

ggsave('../figures/codon_proportion_host_tis.pdf', 
       units = 'in')
```

# Get genes that are induced upon variout treatments.

These are calculated in <analyze_host_fold_changes.md>.
```{r}
induced_genes <- list.files("../tables/", pattern = "induced_genes.tsv", 
                            full.names = T) %>% 
  enframe("sno", "file") %>% 
  mutate(data = map(file, read_tsv)) %>% 
  unnest() %>% 
  select(-sno, -file, -treatment2) %>% 
  rename(sample = treatment1, gene_name = symbol) %>% 
  print()
```

# Plot number of induced genes in each sample

```{r, fig.width=1.8, fig.height=1.6}
plot_data <- induced_genes %>% 
  group_by(sample) %>% 
  summarize(n = n()) %>% 
  mutate(sample = sample_names[sample]) %>% 
  print()

plot_data %>% 
  ggplot(aes(x = sample, y = n)) +
  geom_col(fill = "grey", color = "black") +
  labs(x = "sample", y = "> 2-fold induced\ngene count") 

ggsave('../figures/number_of_induced_host_genes.pdf', 
       units = 'in')
```

Plot overlap between genes induced in +ifn and +ifn+vir sample
```{r, fig.width=1.2, fig.height=1.2}
gene_overlap_ifn_samples <- induced_genes %>% 
  filter(str_detect(sample, "ifn")) %>% 
  group_by(gene_id) %>% 
  mutate(nsamples = n()) %>% 
  mutate(combo = paste0(as.character(sample), collapse = ":")) %>% 
  slice(1) %>% 
  ungroup() %>% 
  group_by(combo) %>% 
  dplyr::count() %>% 
  ungroup() %>% 
  as.data.frame() %>%
  column_to_rownames("combo") %>%
  print()

p <- VennDiagram::draw.pairwise.venn(
  gene_overlap_ifn_samples[['ifn', 'n']] + gene_overlap_ifn_samples[['ifn:ifn_vir', 'n']] ,
  gene_overlap_ifn_samples[['ifn_vir', 'n']] + gene_overlap_ifn_samples[['ifn:ifn_vir', 'n']],
  gene_overlap_ifn_samples[['ifn:ifn_vir', 'n']],
  category = c("+ifn", "+ifn+vir"),
  fill = "grey",
  cat.pos = c(-40, 30),
  cat.dist = c(0.04,0.04))

ggsave("../figures/induced_gene_overlap_between_two_ifn_samples.pdf", plot = p)
```

# Plot the proportion  of different TIS types among induced genes

```{r, fig.width=3, fig.height=1.6}
plot_data <- tis_combo %>% 
  inner_join(induced_genes %>% select(sample, gene_name), by = "gene_name") %>% 
  bind_rows(tis_combo) %>% 
  filter(codon != "other") %>%
  group_by(tis.type, sample) %>% 
  dplyr::count() %>% 
  ungroup() %>% 
  mutate(sample = sample_names[sample]) %>% 
  group_by(sample) %>%
  mutate(n_prop = n / sum(n) * 100) %>%
  ungroup() %>%
  mutate(sample = if_else(is.na(sample), "all", sample)) %>% 
  print(n = 70)

plot_data %>% 
  ggplot(aes(y = n_prop, x = sample, fill = tis.type)) +
  geom_col() +
  geom_text(data = plot_data %>%
              group_by(sample) %>%
              mutate(n = paste0("N=", sum(n))) %>% slice(1), y = 95, aes(label = n), size = 2.8) +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "sample", y = "TIS proportion (%)", 
       fill = "") +
  guides(fill = guide_legend(keyheight = 0.8))

ggsave('../figures/tis_type_induced_genes.pdf', 
       units = 'in')
```


# Test if uTIS or dTIS initiation is overrepresented among the induced genes in comparison with all genes

First get the number of TIS types in the induced genes from the above plot data
```{r}
gene_induction_type_data <- plot_data %>% 
  mutate(tis.type = if_else(tis.type == "aTIS", "aTIS", "altTIS")) %>% 
  group_by(sample, tis.type) %>% 
  summarize(n = sum(n)) %>% 
  ungroup() %>% 
  select(sample, tis.type, n) %>% 
  spread(tis.type, n, fill = 0) %>% 
  data.table::data.table() %>%
  data.table::setkey(sample) %>% 
  print()
```

Test  for increased altTIS in +ifn induced genes
```{r}
data <- gene_induction_type_data[c("+ifn", "all"),c("aTIS", "altTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

Test  for increased altTIS in +ifn +vir induced genes
```{r}
data <- gene_induction_type_data[c("+ifn +vir", "all"),c("aTIS", "altTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

Test for increased altTIS in +vir induced genes
```{r}
data <- gene_induction_type_data[c("+vir", "all"),c("aTIS", "altTIS")] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

# Plot the proportion  of different TIS codons among induced genes

```{r, fig.width=3, fig.height=1.6}
plot_data <- tis_combo %>% 
  inner_join(induced_genes %>% select(sample, gene_name), by = "gene_name") %>% 
  bind_rows(tis_combo) %>% 
  filter(codon != "other") %>%
  group_by(codon, sample) %>% 
  dplyr::count() %>% 
  mutate(codon_priority = tis_priority[codon]) %>% 
  ungroup() %>% 
  mutate(codon = forcats::fct_reorder(factor(codon), codon_priority)) %>% 
  mutate(sample = sample_names[sample]) %>% 
  group_by(sample) %>%
  mutate(n_prop = n / sum(n) * 100) %>%
  ungroup() %>%
  mutate(sample = if_else(is.na(sample), "all", sample))

plot_data %>% 
  ggplot(aes(y = n_prop, x = sample, fill = codon)) +
  geom_col() +
  geom_text(data = plot_data %>%
              group_by(sample) %>%
              mutate(n = paste0("N=", sum(n))) %>% slice(1), y = 95, aes(label = n), size = 2.8) +
  viridis::scale_fill_viridis(discrete = T, direction = -1) +
  labs(x = "sample", y = "TIS proportion (%)", 
       fill = "") +
  guides(fill = guide_legend(keyheight = 0.8))

ggsave('../figures/tis_codon_proportion_induced_genes.pdf', 
       units = 'in')
```

# Test for increased proportion of non-AUG TIS among induced genes

First get the number of TIS codons of different types in the induced genes
```{r}
gene_induction_codon_data <- plot_data %>% 
  mutate(codon = if_else(codon == "AUG", "atg", "nearatg")) %>% 
  group_by(sample, codon) %>% 
  summarize(n = sum(n)) %>% 
  ungroup() %>% 
  spread(codon, n, fill = 0) %>% 
  data.table::data.table() %>% 
  data.table::setkey(sample) %>% 
  print()
```

Test  for increased near AUG in +ifn induced genes
```{r}
data <- gene_induction_codon_data[c("+ifn", "all"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

Test  for increased near AUG in +ifn +vir induced genes
```{r}
data <- gene_induction_codon_data[c("+ifn +vir", "all"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```

Test  for increased near AUG in +vir induced genes
```{r}
data <- gene_induction_codon_data[c("+vir", "all"),!"sample"] %>% 
  as.matrix() 

print(data)
prop.test(data, alternative = "less")
```